### DISCUSSION TITLE
how to create a scalable telegram bot to be used by multiple users simultaneously

### DISCUSSION THREAD

**Discussion by simathedev** [telegraf/telegraf#1721] [Q&A] [answered]
Hello everyone.

I'm an "intermediate rookie"(I hope that makes sense) when it comes to working with node.js.

I am coding a telegram bot using node.js, telegraf and mysql which will be used by 50+ people simultaneously.

The bot has been used by 4 people so far, for testing (but not simultaneously. It's currently making use of long-polling instead of webhooks because I attempted implementing webhooks and failed.

If you have experience coding telegram bots with node.js+telegraf, could you PLEASE give me guidance :(:( I've been stuck with these questions for months:{

Even after research still didn't get many answers.

Could you please answer these questions.

1. If an error occurs with one user, will it affect the other users? example if one user enters the incorrect data which produces an error, will the bot stop for the other users using the bot?

2. How can I make the bot run "forever" especially when an error occurs, how can I handle the error without the bot halting/stopping to the point where I need to restart the bot.

3. I would like to make the bot more scalable. What code structure can I follow to achieve that?

4. One user has complained about the bot repeating a scene and after entering "/start" it works normally. Have you ever experienced? What was your solution.

Your assistance would mean a lot.

**Comment by humblecoder** [comment, marked as answer]
I ended up using a separate "backend" to manage all actual bot activity.  I only used the "bot" proper in order to send/receive interactions.

My latest bots make use of NestJS (not to be confused with NextJS) and my bot simply acts as one of the "services" within the application.

As far as individual users is concerned, you simply manage that from the `ctx` that the bot provides and any information within.

Answer 1:  Appropriate level of `try/catch`, and I simply provide a `heartbeat` endpoint from Nest, that, if something fails within the system, we use Docker (or PM2 if running on system) to restart.

Answer 2:  See Answer 1

Answer 3:  NestJS helped me

Answer 4:  Yeah, this is a weird one.  It's happened, but I use clearly delineated stages and actions as well as internally custom "context" to mitigate this.
